### DISCUSSION TITLE
Typed and Advanced Document and View Settings Class

### DISCUSSION THREAD

**Discussion by GuyGooL5** [BeanieODM/beanie#841] [Feature Request]
One thing that bothers me is the lack of typing and extensibility of the `Settings` class inside documents and views.

For example, in my application I have a custom encoder for `datetime.date`, and to mitigate I created a base document to inherit from, here's an example:

```py
DEFAULT_ENCODERS = {
    datetime.date: lambda x: x.isoformat(),
}


class BaseDocument(Document):
    class Settings:
        bson_encoders = DEFAULT_ENCODERS


class User(BaseDocument):
    email: str
    last_active: datetime.date
    # ...

class Transaction(BaseDocument):
    date: datetime.date
    # ...

```
---
The problems begins when I want to modify the Settings of my class, for example, I want to introduce an Index for email on users:

```py
class User(BaseDocument):
    email: str
    last_active: datetime.date
    # ...

    class Settings:
        indexes = ["email"]
```
By doing so, I have overridden the original Settings class, and thus lost the custom encoders.

After playing around with inheritance and stitching, I came to the conclusion that every method breaks with scale.

After a long thought train I came with the idea of making a producer function for a Settings class.  For starters it can be fully typed, here's an example of the function signature

```py
def DocumentSettings(
    *,
    collection_name: Optional[str] = None,
    indexes: Optional[List[str | Index]] = None,
    bson_encoders: Optional[Dict[type, Callable]] = None,
    keep_null: Optional[bool] = None,
    # ... and the rest of the settings
):
    ...
```

Regarding the implementation, we can create a new class definition that will act as Settings, here's a minimal version of how it might look like

```py
def DocumentSettings(
    *,
    collection_name: Optional[str] = None,
    indexes: Optional[List[str | Index]] = None,
    bson_encoders: Optional[Dict[type, Callable]] = None,
    keep_null: Optional[bool] = None,
    # ... and the rest of the settings
):
    settings_dict = {}
    # now we check for the existence of each setting and add it to the dict
    # for example if collection_name is not None,
    # we add it to the dict as settings_dict["name"] = collection_name

    # ...

    return type("Settings", (), settings_dict)
```

To have it work with `Document`, we can introduce a class field for the model of `Document` with a default value 
```py
class Document(
    LazyModel,
    SettersInterface,
    InheritanceInterface,
    FindInterface,
    AggregateInterface,
    OtherGettersInterface,
):
    Settings: ClassVar[type] = DocumentSettings()
    # ... rest of the implementation
```
---
Now that we have the option to construct the settings class using a function, using the example above

```py
class User(Document):
    email: str
    last_active: datetime.date
    # ...

    Settings = DocumentSettings(
        indexes=["email"],
    )
```
Although we can now create settings in a typed environment, the original problem remains unsolved as we override the whole settings class and lose our encoders yet again.

That's where extensibility comes in. Let's define a base case for DocumentSettings with our custom encoders:
```py
BaseDocumentSettings = DocumentSettings(
    bson_encoders=DEFAULT_ENCODERS,
)
```

Assuming we implement such a method, we can chain or extend an existing settings class that produces a new one.
This is how the implementation would look like for the above example:
```py
class User(Document):
    email: str
    last_active: datetime.date
    # ...

    Settings = BaseDocumentSettings.extend(
        indexes=["email"],
    )
```

This approach makes a non-intrusive reusable and extensible solution.

There are though some setbacks and things to address, like what is the definition of extension for some cases like `keep_nulls`, if a base settings class has `keep_nulls=True` and I extend it with `keep_nulls=False`, what does it mean?
Another thing is omitting, what if I want to remove an existing index or an encoder upon extension?

---

In conclusion, this feature request has a take on two independent features that can be used together:
- type-safe settings
- extensible settings

PS: this whole idea can be easily used on Views, and any other future derivatives as well.
